libraries/botbuilder-adapter-tyntec-whatsapp/src/index.ts

```diff
@@ -0,0 +1,481 @@
+        }
+    }
+
+    public async sendActivities(context: TurnContext, activities: Partial<Activity>[]): Promise<ResourceResponse[]> {
+        const responses: ResourceResponse[] = [];
+        for (const activity of activities) {
+            const tyntecRequest = this.composeTyntecWhatsAppMessageRequest(activity);
+            const axiosRequest = composeTyntecSendWhatsAppMessageRequestConfig(this.tyntecApikey, tyntecRequest);
+            axiosRequest.validateStatus = (): boolean => true;
+
+            const axiosResponse = await this.axiosInstance.request(axiosRequest);
+
+            const messageId = parseTyntecSendWhatsAppMessageResponse(axiosResponse);
+            responses.push({id: messageId});
+        }
+        return responses;
+    }
+
+    public async updateActivity(context: TurnContext, activity: Partial<Activity>): Promise<ResourceResponse | void> {
+        throw Error('Operation updateActivity not supported.');
+    }
+
+    protected composeTyntecWhatsAppMessageRequest(activity: Partial<Activity>): ITyntecWhatsAppMessageRequest {
+        if (activity.type !== ActivityTypes.Message) {
+            throw Error(`TyntecWhatsAppAdapter: Activity.type other than ${ ActivityTypes.Message } not supported: ${ activity.type }`);
+        }
+        if (activity.attachmentLayout !== undefined) {
+            console.warn(`TyntecWhatsAppAdapter: Activity.attachmentLayout not supported: ${ activity.attachmentLayout }`);
+        }
+        if (activity.channelId !== 'whatsapp') {
+            throw Error(`TyntecWhatsAppAdapter: Activity.channelId other than whatsapp not supported: ${ activity.channelId }`);
+        }
+        if (activity.conversation === undefined) {
+            throw Error(`TyntecWhatsAppAdapter: Activity.conversation is required: ${ activity.conversation }`);
+        }
+        if (activity.deliveryMode !== undefined) {
+            throw Error(`TyntecWhatsAppAdapter: Activity.deliveryMode not supported: ${ activity.deliveryMode }`);
+        }
+        if (activity.entities !== undefined) {
+            console.warn(`TyntecWhatsAppAdapter: Activity.entities not supported: ${ activity.entities }`);
+        }
+        if (activity.expiration !== undefined) {
+            console.warn(`TyntecWhatsAppAdapter: Activity.expiration not supported: ${ activity.expiration }`);
+        }
+        if (activity.from === undefined) {
+            throw Error(`TyntecWhatsAppAdapter: Activity.from is required: ${ activity.from }`);
+        }
+        if (activity.id !== undefined) {
+            console.warn(`TyntecWhatsAppAdapter: Activity.id not supported: ${ activity.id }`);
+        }
+        if (activity.importance !== undefined) {
+            console.warn(`TyntecWhatsAppAdapter: Activity.importance not supported: ${ activity.importance }`);
+        }
+        if (activity.inputHint !== undefined) {
+            throw Error(`TyntecWhatsAppAdapter: Activity.inputHint not supported: ${ activity.inputHint }`);
```

Comment:
Should the adapter throw errors on unsupported features? People tend to use the Bot Framework SDK to build bots for multiple channels, where the Bot Service or adapter will take care of this. Throwing errors will cause the bot to stop working, where I believe that [graceful degradation / progressive enhancement](https://www.w3.org/wiki/Graceful_degradation_versus_progressive_enhancement) would be a better strategy. Just pass through what is supported and ignore unsupported features.

Reply:
I don't know. Is there a best practice? It throws an error only in few cases where it seemed to me that it might confuse the user if the adapter only ignored that properties. If you think that a warning is enough, I'll change it.

Reply:
I've changed the adapter to throw a warning for `inputHint` in order to let it support `MessageFactory.text` and `sendActivity(string)`. But I think that your comment applies to the other errors as well. Please see my question above.